### PR TITLE
feat(plasma-web): add a11y for date picker (Calendar)

### PR DESCRIPTION
**Pull request by Yakutoc**
Добавить поддержку доступности согласно [примеру(Date Picker Dialog Example)](https://www.w3.org/WAI/ARIA/apg/example-index/dialog-modal/datepicker-dialog.html)

Проверить на доступность в:

macOS

[VoiceOver](https://www.apple.com/ru/accessibility/vision/) - (нативные возможности системы);
Windows

[JAWS](https://www.freedomscientific.com/products/software/jaws/) (платный[ограничения в 40 мин пользования]);
[NVDA](https://www.nvaccess.org/about-nvda/) (бесплатный);

:baby_chick: Download canary assets:

[colors--canary.222.3522295576.xml](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/colors--canary.222.3522295576.xml)  
[color_sbermarket_ios-swift--canary.222.3522295576.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_ios-swift--canary.222.3522295576.swift)  
[color_sbermarket_kotlin--canary.222.3522295576.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_kotlin--canary.222.3522295576.kt)  
[color_sbermarket_react-native--canary.222.3522295576.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_react-native--canary.222.3522295576.ts)  
[color_sberprime_ios-swift--canary.222.3522295576.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_ios-swift--canary.222.3522295576.swift)  
[color_sberprime_kotlin--canary.222.3522295576.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_kotlin--canary.222.3522295576.kt)  
[color_sberprime_react-native--canary.222.3522295576.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_react-native--canary.222.3522295576.ts)  
[PlasmaTokensColor--canary.222.3522295576.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/PlasmaTokensColor--canary.222.3522295576.swift)  
[typo_mage_kotlin--canary.222.3522295576.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_kotlin--canary.222.3522295576.kt)  
[typo_mage_react-native--canary.222.3522295576.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_react-native--canary.222.3522295576.ts)  
[typo_plasma_kotlin--canary.222.3522295576.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_kotlin--canary.222.3522295576.kt)  
[typo_plasma_react-native--canary.222.3522295576.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_react-native--canary.222.3522295576.ts)  
[typo_ruler_kotlin--canary.222.3522295576.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_kotlin--canary.222.3522295576.kt)  
[typo_ruler_react-native--canary.222.3522295576.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_react-native--canary.222.3522295576.ts)  
[typo_sage_kotlin--canary.222.3522295576.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_kotlin--canary.222.3522295576.kt)  
[typo_sage_react-native--canary.222.3522295576.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_react-native--canary.222.3522295576.ts)  
[typo_sbermarket_kotlin--canary.222.3522295576.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_kotlin--canary.222.3522295576.kt)  
[typo_sbermarket_react-native--canary.222.3522295576.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_react-native--canary.222.3522295576.ts)  
[typo_soulmate_kotlin--canary.222.3522295576.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_kotlin--canary.222.3522295576.kt)  
[typo_soulmate_react-native--canary.222.3522295576.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_react-native--canary.222.3522295576.ts)

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @salutejs/plasma-b2c@1.115.0-canary.222.3522295576.0
  npm install @salutejs/plasma-icons@1.109.0-canary.222.3522295576.0
  npm install @salutejs/plasma-temple@1.113.0-canary.222.3522295576.0
  npm install @salutejs/plasma-ui@1.146.0-canary.222.3522295576.0
  npm install @salutejs/plasma-web@1.143.0-canary.222.3522295576.0
  # or 
  yarn add @salutejs/plasma-b2c@1.115.0-canary.222.3522295576.0
  yarn add @salutejs/plasma-icons@1.109.0-canary.222.3522295576.0
  yarn add @salutejs/plasma-temple@1.113.0-canary.222.3522295576.0
  yarn add @salutejs/plasma-ui@1.146.0-canary.222.3522295576.0
  yarn add @salutejs/plasma-web@1.143.0-canary.222.3522295576.0
  ```
</details>

### CODE REVIEW

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-222/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-222/
  ui storybook: http://plasma.sberdevices.ru/pr-222/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-222/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-222/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-222/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-222/
  <details>
    <summary>New documentation</summary>
    <br />

    http://plasma.fine.sberdevices.ru/pr/demo-canvas-app-pr-222/

  </details>

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-222/
  ui storybook: http://plasma.sberdevices.ru/pr-222/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-222/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-222/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-222/temple-storybook/
  <details>
    <summary>New documentation</summary>
    <br />

    website: http://plasma.fine.sberdevices.ru/pr/pr-222/
    ui storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/ui-storybook/
    web storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/web-storybook/
    b2c storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/b2c-storybook/
    temple storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/temple-storybook/

  </details>

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-222/
  <details>
    <summary>New documentation</summary>
    <br />

    http://plasma.fine.sberdevices.ru/pr/demo-canvas-app-pr-222/

  </details>

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-222/
  ui storybook: http://plasma.sberdevices.ru/pr-222/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-222/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-222/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-222/temple-storybook/
  <details>
    <summary>New documentation</summary>
    <br />

    website: http://plasma.fine.sberdevices.ru/pr/pr-222/
    ui storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/ui-storybook/
    web storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/web-storybook/
    b2c storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/b2c-storybook/
    temple storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/temple-storybook/

  </details>

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-222/
  <details>
    <summary>New documentation</summary>
    <br />

    http://plasma.fine.sberdevices.ru/pr/demo-canvas-app-pr-222/

  </details>

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-222/
  ui storybook: http://plasma.sberdevices.ru/pr-222/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-222/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-222/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-222/temple-storybook/
  <details>
    <summary>New documentation</summary>
    <br />

    website: http://plasma.fine.sberdevices.ru/pr/pr-222/
    ui storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/ui-storybook/
    web storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/web-storybook/
    b2c storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/b2c-storybook/
    temple storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/temple-storybook/

  </details>

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-222/
  <details>
    <summary>New documentation</summary>
    <br />

    http://plasma.fine.sberdevices.ru/pr/demo-canvas-app-pr-222/

  </details>

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-222/
  <details>
    <summary>New documentation</summary>
    <br />

    http://plasma.fine.sberdevices.ru/pr/demo-canvas-app-pr-222/

  </details>

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-222/
  ui storybook: http://plasma.sberdevices.ru/pr-222/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-222/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-222/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-222/temple-storybook/
  <details>
    <summary>New documentation</summary>
    <br />

    website: http://plasma.fine.sberdevices.ru/pr/pr-222/
    ui storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/ui-storybook/
    web storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/web-storybook/
    b2c storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/b2c-storybook/
    temple storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/temple-storybook/

  </details>

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-222/
  <details>
    <summary>New documentation</summary>
    <br />

    http://plasma.fine.sberdevices.ru/pr/demo-canvas-app-pr-222/

  </details>

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-222/
  ui storybook: http://plasma.sberdevices.ru/pr-222/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-222/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-222/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-222/temple-storybook/
  <details>
    <summary>New documentation</summary>
    <br />

    website: http://plasma.fine.sberdevices.ru/pr/pr-222/
    ui storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/ui-storybook/
    web storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/web-storybook/
    b2c storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/b2c-storybook/
    temple storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/temple-storybook/

  </details>

**Review by Yeti-or — APPROVED** [review]

File: packages/plasma-icons/src/IconRoot.tsx [hunk 1]
```diff
@@ -39,7 +39,7 @@ export const IconRoot: React.FC<IconRootProps> = ({ icon: IconComponent, size, c
     const w = `${sizeMap[size]}rem`;
 
     return (
-        <StyledRoot w={w} className={className}>
+        <StyledRoot aria-hidden w={w} className={className}>
```
[the remark this answers is not in the document]
Comment (Yakutoc) [reply packages/plasma-icons/src/IconRoot.tsx:42]: @Yeti-or Это иконка. Они не несут никакой смысловой нагрузки в aria дереве. А вот ридеры их читаю. 

Например в случае с кнопкой предыдущий месяц, ридер прочтет текст, кнопка, группа (это как иконка). 

А с включенным  aria-hidden="true" этого не будет.

File: packages/plasma-web/src/components/Calendar/CalendarBase.tsx [hunk 2]
```diff
@@ -52,37 +52,57 @@ export const CalendarBase: React.FC<CalendarBaseProps> = ({
...
-            dispatch({ type: ActionType.NEXT_START_YEAR, payload: { yearsCount: YEAR_RENDER_COUNT } });
-        }
-    }, [date, calendarState]);
+    const handlePrev = useCallback(
+        (withShift = false) => {
+            if (calendarState === CalendarState.Days) {
+                const data = {
+                    type: withShift ? ActionType.PREVIOUS_YEAR : ActionType.PREVIOUS_MONTH,
+                    payload: withShift ? { step: 1 } : { monthIndex: date.monthIndex, year: date.year },
+                } as PreviousYearAction | PreviousMonthAction;
```
Comment (neretin-trike) [inline packages/plasma-web/src/components/Calendar/CalendarBase.tsx:61]: а для чего тут явное приведение типов нужно?

File: packages/plasma-web/src/components/Calendar/CalendarBase.tsx [hunk 3]
```diff
@@ -52,37 +52,57 @@ export const CalendarBase: React.FC<CalendarBaseProps> = ({
...
+        [date, calendarState],
+    );
+
+    const handleNext = useCallback(
+        (withShift = false) => {
+            if (calendarState === CalendarState.Days) {
+                const data = {
+                    type: withShift ? ActionType.NEXT_YEAR : ActionType.NEXT_MONTH,
+                    payload: withShift ? { step: 1 } : { monthIndex: date.monthIndex, year: date.year },
+                } as NextMonthAction | NextYearAction;
```
Comment (neretin-trike) [inline packages/plasma-web/src/components/Calendar/CalendarBase.tsx:87]: и здесь

File: packages/plasma-web/src/components/Calendar/CalendarDayItem.tsx [hunk 4]
```diff
@@ -171,9 +172,11 @@ export const CalendarDayItem = memo(
                     data-day={day}
                     data-month-index={monthIndex}
                     data-year={year}
+                    aria-selected={isSelected}
+                    {...rest}
```
Comment (neretin-trike) [inline packages/plasma-web/src/components/Calendar/CalendarDayItem.tsx:176]: почему rest в итоге удалил?

Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/CalendarDayItem.tsx:176]: @neretin-trike не понял вопроса.

File: packages/plasma-web/src/components/Calendar/CalendarBase.tsx (hunk 2 above)
Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/CalendarBase.tsx:61]: @neretin-trike Пришлось так сделать иначе TS сходил с ума.

```
TS2345: Argument of type '{ type: ActionType; payload: { step: number; monthIndex?: undefined; year?: undefined; } | { monthIndex: number; year: number; step?: undefined; }; }' is not assignable to parameter of type 'Action'.   Type '{ type: ActionType; payload: { step: number; monthIndex?: undefined; year?: undefined; } | { monthIndex: number; year: number; step?: undefined; }; }' is not assignable to type 'UpdateCalendarStateAction'.     Types of property 'type' are incompatible.       Type 'ActionType' is not assignable to type 'ActionType.UPDATE_CALENDAR_STATE'.
```

Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/CalendarBase.tsx:61]: После обсуждения решили исправить. 

Делаем в таком формате 

```js
if (calendarState === CalendarState.Days) {
      if (withShift) {
             dispatch({
                    type: ActionType.PREVIOUS_YEAR,
                   payload: { step: 1 },
              });

                  return;
            }

                dispatch({
                    type: ActionType.PREVIOUS_MONTH,
                    payload: { monthIndex: date.monthIndex, year: date.year },
                });

                return;
}
```

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-222/
  <details>
    <summary>New documentation</summary>
    <br />

    http://plasma.fine.sberdevices.ru/pr/demo-canvas-app-pr-222/

  </details>

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-222/
  ui storybook: http://plasma.sberdevices.ru/pr-222/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-222/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-222/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-222/temple-storybook/
  <details>
    <summary>New documentation</summary>
    <br />

    website: http://plasma.fine.sberdevices.ru/pr/pr-222/
    ui storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/ui-storybook/
    web storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/web-storybook/
    b2c storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/b2c-storybook/
    temple storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/temple-storybook/

  </details>

**Review by neretin-trike — APPROVED** [review]

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-222/
  <details>
    <summary>New documentation</summary>
    <br />

    http://plasma.fine.sberdevices.ru/pr/demo-canvas-app-pr-222/

  </details>

**Review by neretin-trike — APPROVED** [review]

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-222/
  <details>
    <summary>New documentation</summary>
    <br />

    http://plasma.fine.sberdevices.ru/pr/demo-canvas-app-pr-222/

  </details>

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-222/
  ui storybook: http://plasma.sberdevices.ru/pr-222/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-222/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-222/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-222/temple-storybook/
  <details>
    <summary>New documentation</summary>
    <br />

    website: http://plasma.fine.sberdevices.ru/pr/pr-222/
    ui storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/ui-storybook/
    web storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/web-storybook/
    b2c storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/b2c-storybook/
    temple storybook: http://plasma.fine.sberdevices.ru/pr/pr-222/temple-storybook/

  </details>

**Pull request merged**